### DISCUSSION TITLE
Explicit valence for atom # 0 I, 7, is greater than permitted

### DISCUSSION THREAD

**Discussion by teltim** [rdkit/rdkit#6555] [FAQ] [answered]
I tested IF7 (https://pubchem.ncbi.nlm.nih.gov/compound/85645) with the following 

suppl = Chem.SDMolSupplier("sdf_files/Structure2D_COMPOUND_CID_85645.sdf", removeHs=False)
smile = "FI(F)(F)(F)(F)(F)F"

[11:48:05] Explicit valence for atom # 0 I, 7, is greater than permitted
[11:48:05] ERROR: Could not sanitize molecule ending on line 20
[11:48:05] ERROR: Explicit valence for atom # 0 I, 7, is greater than permitted

BrF5

suppl = Chem.SDMolSupplier("sdf_files/Structure2D_COMPOUND_CID_24606.sdf", removeHs=False)
smile = "FBr(F)(F)(F)F"
[15:52:34] Explicit valence for atom # 0 Br, 5, is greater than permitted
[15:52:34] ERROR: Could not sanitize molecule ending on line 16
[15:52:34] ERROR: Explicit valence for atom # 0 Br, 5, is greater than permitted

ClF5

suppl = Chem.SDMolSupplier("sdf_files/Structure2D_COMPOUND_CID_61654.sdf", removeHs=False)
smile = "FCl(F)(F)(F)F"
[16:02:50] Explicit valence for atom # 0 Cl, 5, is greater than permitted
[16:02:50] ERROR: Could not sanitize molecule ending on line 16
[16:02:50] ERROR: Explicit valence for atom # 0 Cl, 5, is greater than permitted

Are there any countermeasures for those moleclues

**Comment by teltim** [comment]
I need some advice

**Comment by ptosco** [comment]
You may switch off sanitization in `SDMolSupplier` and sanitize manually, catching sanitize exceptions such as the one you are incurring into:
```
from rdkit import Chem

suppl = Chem.SDMolSupplier("Structure2D_COMPOUND_CID_85645.sdf", removeHs=False, sanitize=False)
mol = next(suppl)
try:
    Chem.SanitizeMol(mol)
except Chem.AtomValenceException:
    # here you may handle the error condition, or just ignore it
    pass
mol
```
![image](https://github.com/rdkit/rdkit/assets/5244385/50e85132-0656-4243-b0e4-446b8c2cad96)
Please note that you will incur into some limitations in what you can do with non-sanitized molecules as all operations that involve valence calculations will fail.

**Reply by teltim** [reply]
Thank you for the kind advice and comments, it is very helpfull for me.
I'm actually beginner in chemistry, physical chemistry, and quantum chemisty...

As for "limitations in what you can do with non-sanitized molecules" 
Could you teach me, as example, what kinds of calculation I can not perform with "non-sanitized molecules" ??

Best regards!

**Reply by ptosco** [reply]
For example, attempting to add explicit Hs with `Chem.AddHs()`, computing an MCS, attempting to assign stereochemistry will fail with
```
Pre-condition Violation
getNumImplicitHs() called without preceding call to calcImplicitValence()
```

**Reply by teltim** [reply]
Thanks for the kind answer, I understand about it.
I try calculation for what I want to do.
best regards

**Reply by teltim** [reply]
Hi, I have another question I'd like to ask...

I did the following, which is Optimization of the 3D structure of the Mol object, for the Electronic structure calculations in the later stages

EmbedMolecule(mol, params)
or
MMFFOptimizeMolecule(mol)

This seems not to be acceptable. 
****
Pre-condition Violation
getNumImplicitHs() called without preceding call to calcImplicitValence()
Violation occurred on line 297 in file /project/build/temp.linux-x86_64-cpython-39/rdkit/Code/GraphMol/Atom.cpp
Failed Expression: d_implicitValence > -1
****

Is this correct?
It means the structure optimization of the moleclues above is not possible w/ RDkit?

I might use other application for the purpose (the structure optimization) above?

**Comment by ptosco** [comment, marked as answer]
Correct. You need valence counts for those. Also, the MMFF forcefields has no parameters for heptavalent iodine. I would just get a reasonable set of starting 3D coordinates (you may find many on the web) and optimise those with a QM method.

**Reply by teltim** [reply]
thank you for the reply
